website/src/components/McapRecordingDemo/videoCapture.ts

```diff
@@ -66,36 +85,106 @@ export function startVideoCapture(params: VideoCaptureParams): () => void {
+  const { video, onJpegFrame, onH264Frame, onError, frameDurationSec } = params;
   const canvas = document.createElement("canvas");
   canvas.width = video.videoWidth;
   canvas.height = video.videoHeight;
   const ctx = canvas.getContext("2d");
 
-  let framePromise: Promise<void> | undefined;
+  let encoder: VideoEncoder | undefined;
+  const framePool: ArrayBuffer[] = [];
+  setupEncoder: try {
+    const config: VideoEncoderConfig = {
+      codec: "avc1.42001f", // Baseline profile (42 00) with maximum level (1f)
+      width: video.videoWidth,
+      height: video.videoHeight,
+      latencyMode: "realtime",
+      avc: { format: "annexb" },
+    };
+    if (typeof VideoEncoder !== "function") {
+      break setupEncoder;
+    }
+    if ((await VideoEncoder.isConfigSupported(config)).supported !== true) {
+      break setupEncoder;
+    }
+    encoder = new VideoEncoder({
+      output: (chunk) => {
+        if (signal.aborted) {
+          return;
+        }
+        let buffer = framePool.pop();
+        if (!buffer || buffer.byteLength < chunk.byteLength) {
+          buffer = new ArrayBuffer(chunk.byteLength);
+        }
+        chunk.copyTo(buffer);
+        onH264Frame({
+          data: new Uint8Array(buffer, 0, chunk.byteLength),
+          release() {
+            framePool.push(this.data.buffer);
+          },
+        });
+      },
+      error: onError,
+    });
+
+    encoder.configure(config); // may throw
+  } catch (err) {
+    onError(err as Error);
+    encoder = undefined;
+  }
+
+  // add a keyframe every 2 seconds for h264 encoding
+  const keyframeInterval = 2000;
+  let lastKeyframeTime: number | undefined;
+
+  let processingFrame = false;
   const interval = setInterval(() => {
-    if (framePromise) {
+    if (processingFrame) {
       // last frame is not yet complete, skip frame
       return;
     }
-    framePromise = new Promise((resolve) => {
+    processingFrame = true;
+    if (encoder) {
+      encoder.addEventListener(
+        "dequeue",
+        () => {
+          processingFrame = false;
+        },
+        { once: true },
+      );
+      const now = performance.now();
+      const frame = new VideoFrame(video);
+      const encodeOptions: VideoEncoderEncodeOptions = {};
+      if (
+        lastKeyframeTime == undefined ||
+        now - lastKeyframeTime >= keyframeInterval
+      ) {
+        encodeOptions.keyFrame = true;
+        lastKeyframeTime = now;
+      }
+      encoder.encode(frame, encodeOptions);
+      frame.close();
+    } else {
       ctx?.drawImage(video, 0, 0);
       canvas.toBlob(
         (blob) => {
+          processingFrame = false;
           if (blob && !signal.aborted) {
-            onFrame(blob);
+            onJpegFrame(blob);
           }
-          resolve();
-          framePromise = undefined;
         },
         "image/jpeg",
         0.8,
       );
-    });
+    }
   }, frameDurationSec * 1000);
 
   const cleanup = () => {
```

Comment:
Should you do anything to clean up the encoder, event listeners, or framepool here?